Tests/SwiftDiagnosticsTest/GroupDiagnosticsFormatterTests.swift

```diff
@@ -0,0 +1,205 @@
+        ╭-── #myAssert────────────────────────────────────────────────────────
+        │1 │ let __a = pi
+        │2 │ let __b = 3
+        │3 │ if !(__a == __b) {
+        │  ∣          ╰─ no matching operator '==' for types 'Double' and 'Int'
+        │4 │   fatalError("assertion failed: pi != 3")
+        │5 │ }
+        ╰-────────────────────────────────────────────────────────────────────
+      3 │ print("hello"
+        ∣              ╰─ expected ')' to end function call
+
+      """
+    )
+  }
+
+  func testGroupingForDoubleNestedMacroExpansion() {
+    var group = GroupedDiagnostics()
+
+    // Main source file.
+    let (mainSourceID, mainSourceMarkers) = group.addTestFile(
+      """
+      let pi = 3.14159
+      0️⃣#myAssert(pi == 3)
+      print("hello"
+      """,
+      displayName: "main.swift",
+      extraDiagnostics: ["0️⃣": ("in expansion of macro 'myAssert' here", .note)]
+    )
+    let inExpansionNotePos = mainSourceMarkers["0️⃣"]!
+
+    // Outer expansion source file
+    let (outerExpansionSourceID, outerExpansionSourceMarkers) = group.addTestFile(
+      """
+      let __a = pi
+      let __b = 3
+      if 0️⃣#invertedEqualityCheck(__a, __b) {
+        fatalError("assertion failed: pi != 3")
+      }
+      """,
+      displayName: "#myAssert",
+      parent: (mainSourceID, inExpansionNotePos),
+      extraDiagnostics: [
+        "0️⃣": ("in expansion of macro 'invertedEqualityCheck' here", .note)
+      ]
+    )
+    let inInnerExpansionNotePos = outerExpansionSourceMarkers["0️⃣"]!
+
+    // Expansion source file
+    _ = group.addTestFile(
+      """
+      !(__a 0️⃣== __b)
+      """,
+      displayName: "#invertedEqualityCheck",
+      parent: (outerExpansionSourceID, inInnerExpansionNotePos),
+      extraDiagnostics: [
+        "0️⃣": ("no matching operator '==' for types 'Double' and 'Int'", .error)
+      ]
+    )
+
+    let formatter = DiagnosticsFormatter()
+    let annotated = formatter.annotateSources(in: group)
+    print(annotated)
+    AssertStringsEqualWithDiff(
+      annotated,
+      """
+      === main.swift ===
+      1 │ let pi = 3.14159
+      2 │ #myAssert(pi == 3)
+        ∣ ╰─ in expansion of macro 'myAssert' here
+        ╭-── #myAssert────────────────────────────────────────────────────────
+        │1 │ let __a = pi
+        │2 │ let __b = 3
+        │3 │ if #invertedEqualityCheck(__a, __b) {
+        │  ∣    ╰─ in expansion of macro 'invertedEqualityCheck' here
+        │  ╭-── #invertedEqualityCheck────────────────────────────────────────
+        │  │1 │ !(__a == __b)
+        │  │  ∣       ╰─ no matching operator '==' for types 'Double' and 'Int'
+        │  ╰-─────────────────────────────────────────────────────────────────
+        │4 │   fatalError("assertion failed: pi != 3")
+        │5 │ }
+        ╰-────────────────────────────────────────────────────────────────────
+      3 │ print("hello"
+        ∣              ╰─ expected ')' to end function call
+
+      """
```

Comment:
The debug output looks a little bit too squashed to me. What do you think about adding an empty l line before and after starting a new source file buffer and adda leading space after the `|` of the inline source buffer. I.e.

```
      === main.swift ===
       1 │ let pi = 3.14159
       2 │ #myAssert(pi == 3)
         ∣ ╰─ in expansion of macro 'myAssert' here

         ╭-── #myAssert────────────────────────────────────────────────────────
         │ 1 │ let __a = pi
         │ 2 │ let __b = 3
         │ 3 │ if #invertedEqualityCheck(__a, __b) {
         │   ∣    ╰─ in expansion of macro 'invertedEqualityCheck' here
         │ 
         │   ╭-── #invertedEqualityCheck────────────────────────────────────────
         │   │ 1 │ !(__a == __b)
         │   │   ∣       ╰─ no matching operator '==' for types 'Double' and 'Int'
         │   ╰-─────────────────────────────────────────────────────────────────
         │ 
         │ 4 │   fatalError("assertion failed: pi != 3")
         │ 5 │ }
         ╰-────────────────────────────────────────────────────────────────────

       3 │ print("hello"
         ∣              ╰─ expected ')' to end function call

```

Reply:
This is an easy change we can do at any point, but... this looks way too spaced out for me.  In the colorized version, I've made the boxes and line numbers cyan, which helps distinguish things more.

Reply:
Ah, I didn’t look at the colorized output yet. If you’ve got a preference, let’s keep it as-is for now.